demo_apps/whatsapp_llama2.md

```diff
@@ -0,0 +1,151 @@
+pip install langchain replicate flask requests uvicorn gunicorn
+```
+
+Then, create a Python file named llama-chatbot.py with the following code, which defines a class `WhatsAppClient` and a method `send_text_message` to post a message (the answer generated by Llama 2 on a user query) to the WhatsApp Cloud API, which then sends the answer back to the WhatsApp user. Remember to set `WHATSAPP_API_TOKEN` and `WHATSAPP_CLOUD_NUMBER_ID` to the values you saved in the previous section.
+
+```
+import langchain
+from langchain.llms import Replicate
+from flask import Flask
+from flask import request
+import os
+import requests
+import json
+
+class WhatsAppClient:
+    API_URL = "https://graph.facebook.com/v17.0/"
+    WHATSAPP_API_TOKEN = "<Temporary access token from your WhatsApp API Setup>"
+    WHATSAPP_CLOUD_NUMBER_ID = "<Phone number ID from your WhatsApp API Setup>"
+
+    def __init__(self):
+        self.headers = {
+            "Authorization": f"Bearer {self.WHATSAPP_API_TOKEN}",
+            "Content-Type": "application/json",
+        }
+        self.API_URL = self.API_URL + self.WHATSAPP_CLOUD_NUMBER_ID
+        
+    def send_text_message(self, message, phone_number):
+        payload = {
+            "messaging_product": 'whatsapp',
+            "to": phone_number,
+            "type": "text",
+            "text": {
+                "preview_url": False,
+                "body": message
+            }
+        }
+        response = requests.post(f"{self.API_URL}/messages", json=payload, headers=self.headers)
+        return response.status_code
+```
+
+Finally, add the code below to llama-chatbot.py, which creates a Llama 2 instance and defines an HTTP method `msgrcvd` to:
+1. receive the user message forwarded by the webhook;
+2. ask Llama 2 for the answer;
+3. call the `WhatsAppClient`'s `send_text_message`` with a recipient's phone number.
+   
+```   
+os.environ["REPLICATE_API_TOKEN"] = "<your replicate api token>"    
+llama2_13b_chat = "meta/llama-2-13b-chat:f4e2de70d66816a838a89eeeb621910adffb0dd0baba3976c96980970978018d"
+
+llm = Replicate(
+    model=llama2_13b_chat,
+    model_kwargs={"temperature": 0.01, "top_p": 1, "max_new_tokens":500}
+)
+client = WhatsAppClient()
+app = Flask(__name__)
+
+@app.route("/")
+def hello_llama():
+    return "<p>Hello Llama 2</p>"
+
+@app.route('/msgrcvd', methods=['POST', 'GET'])
+def msgrcvd():    
+    message = request.args.get('message')
+    answer = llm(message)
+    client.send_text_message(llm(message), "<a recipient phone number from your WhatsApp API Setup>")
+    return message + "<p/>" + answer
+```
+
+Now it's time to modify the webhook to complete the whole app.
+
+## Modifying the Webhook 
+
+Open your glitch.com webhook URL created earlier, and after the code snippet:
+
+```
+// message received! 
+console.log(req.body["entry"][0]["changes"][0]["value"]["messages"][0]["text"]["body"]);
+```
+
+add the code below - remember to change <your web server IP>:
```

Comment:
Before this step, the server IP needs to be publicly visible, right? Might be good to point this out in the guide, just to cover make users aware of the need for a public service running the app.

Reply:
yes the server IP needs to be publicly visible, before or after this step (but before the last step "open your WhatsApp app, enter a question"). Updated based on your feedback.